File: functions/functions_k.R
# What this file is ----------------------------------------------------------
#
# The k-factor functions, used to put a confidence interval around an
# estimated failure rate. They come in the same four flavors R uses for any
# distribution: qk() for quantiles, pk() for cumulative probabilities, dk()
# for densities, and rk() for random draws.
#
# Load it, from the top of the project folder:
#   source("functions/functions_k.R")
#
# Then try:
#   qk(p = 0.95, r = 20, .time = FALSE, .failure = FALSE)
#
# You never need to edit anything below.
#

#' @name qk
#' @title k-factor Quantiles
#' @description 
#' Function to return quantiles for k-factors.
#' Intended for estimating confidence intervals for failure rates.
#' @param p:[dbl] vector of probabilities / percentile(s)
#' @param r:int number of failures (non-negative integers; can include zero)
#' @param .time:logical logical; is case time-censored data?
#' @param .failure:logical logical; is case failure-censored data?
#' 
#' @importFrom dplyr `case_when`
qk = function(p, r, .time = FALSE, .failure = FALSE){
  # Testing values
  # p = 0.95; r = 20; .time = FALSE; .failure = FALSE
  
  # Input error handling
  stopifnot(is.numeric(p) & p >= 0 & p <= 1)
  stopifnot(is.numeric(as.integer(r)))
  stopifnot(is.logical(.time))
  stopifnot(is.logical(.failure))
  stopifnot(
    (.time == TRUE & .failure == FALSE) | 
      (.time == FALSE & .failure == FALSE) | 
      (.time == FALSE & .failure == TRUE) )
  stopifnot(
    # R should either be greater than 0
    (r > 0) |
      # OR 
      # zero and the data should be time censored
    (r == 0 & .time == TRUE)
  )
  
  # Evaluate if p is in the upper or lower tail
  .upper = p > 0.5
  
  # Does r == 0?
  .zerofailures = r == 0
  
  
  k = case_when(
    # 1+ failures AND complete data AND UPPER tail  --> Get k-factor for r as normal
    .zerofailures == FALSE & .time == FALSE & .failure == FALSE & .upper == TRUE ~ qchisq(p, df = 2*r) / (2*r),
    # 1+ failures AND complete data AND LOWER tail  --> Get k-factor for r as normal
    .zerofailures == FALSE & .time == FALSE & .failure == FALSE & .upper == FALSE ~ qchisq(p, df = 2*r) / (2*r),
    # 1+ failures AND time-censored data AND UPPER tail --> Get k-factor for r+1
    .zerofailures == FALSE & .time == TRUE & .failure == FALSE & .upper == TRUE ~ qchisq(p, df = 2*(r + 1) ) / (2*r),
    # 1+ failures AND time-censored data AND LOWER tail --> Get k-factor for r as normal
    .zerofailures == FALSE & .time == TRUE & .failure == FALSE & .upper == FALSE ~ qchisq(p, df = 2*(r) ) / (2*r),
    
    # 1+ failures AND failure-censored (Type II) data AND UPPER tail
    # --> The 3-factor rule qchisq(p, 2*((r-1)+1)) / (2*(r-1)) * (r-1)/r
    #     reduces algebraically to the complete-data formula, so use it directly.
    .zerofailures == FALSE & .time == FALSE & .failure == TRUE & .upper == TRUE ~ qchisq(p, df = 2*r) / (2*r),
    # 1+ failures AND failure-censored (Type II) data AND LOWER tail --> complete-data formula
    .zerofailures == FALSE & .time == FALSE & .failure == TRUE & .upper == FALSE ~ qchisq(p, df = 2*r) / (2*r),
    
    # If zero failures --> then time-censored --> time = TRUE, and upper/lower distinction doesn't matter.
    .zerofailures == TRUE & .time == TRUE & .failure == FALSE ~ -log(1-p),
    # Otherwise, return NA.
    TRUE ~ NA_real_
  )
  
  if(any(is.na(k))){ message("At least 1 k-factor could not be calculated, due to improper inputs. Review the rules for time-censored, failure-censored, and zero-failure data.")}
    
  return(k) 
}

#' @name rk
#' @title k-factor Random Deviates
#' @description 
#' Get a random sample of k-factor values for simulating sampling distributions of failure rates.
#' @param n:int number of observations.
#' @param r:int number of failures (non-negative integers; can include zero)
#' @param .time:logical logical; is case time-censored data?
#' @param .failure:logical logical; is case failure-censored data?
rk = function(n, r, .time = FALSE, .failure = FALSE){
  
  # Testing values
  # n = 100; r = 20; .time = FALSE; .failure = FALSE
  
  # Input error handling
  stopifnot(is.integer(as.integer(n)) & as.integer(n) > 0)
  stopifnot(is.logical(.time))
  stopifnot(is.logical(.failure))
  stopifnot(
    (.time == TRUE & .failure == FALSE) | 
      (.time == FALSE & .failure == FALSE) | 
      (.time == FALSE & .failure == TRUE) )
  
  # Does r == 0?
  .zerofailures = r == 0
  
  # Generate a uniform distribution of percentiles p
  p_uniform = runif(n = n, min = 0, max = 1)
  
  # Return quantiles for the random percentiles
  k = qk(p = p_uniform, r = r, .time = .time, .failure = .failure)
  return(k)
}

# rk(n = 1000, r = 20) %>% hist()


#' @name pk
#' @title k-factor Cumulative Distribution Function
#' @description 
#' Function to return cumulative probabilities / percentiles given a supplied k-factor quantile `q`.
#' Intended for confidence intervals for failure rates.
#' @param q:[dbl] vector of quantiles (k-factors)
#' @param r:int number of failures (non-negative integers; can include zero)
#' @param .time:logical logical; is case time-censored data?
#' @param .failure:logical logical; is case failure-censored data?
pk = function(q, r, .time = FALSE, .failure = FALSE){
  # Testing values
  # q = 2; r = 20; .time = FALSE; .failure = FALSE
  # We just need to map the function...

  # Construct an approximation function f, which gives the inverse of the Quantile Function,
  # such that you use linear interpolation to return a Probability for any Quantile supplied.
  by = 0.001
  p_range = c(by/10000, by/1000, by/100, by/10, 
              seq(from = 0, to = 1, by = 0.001),
              1 - by/10, 1 - by/100, 1 - by/1000, 1 - by/10000)
  p_range = sort(p_range)
  # Get the quantiles for that range
  q_range = qk(p = p_range, r = r, .time = .time, .failure = .failure)
  # Get the inverse quantile function
  f = approxfun(x = q_range, y = p_range, method = "linear", rule = 2, na.rm = TRUE)
  # Return the expected CDF for that quantile  
  p = f(q)
  return(p)
}


#' @name dk
#' @title k-factor Probability Density Function
#' @description 
#' Function to return probability densities given a supplied k-factor quantile `q`.
#' Intended for visualizing sampling distributions of failure rates.
#' @param x:[dbl] vector of quantiles (k-factors)
#' @param r:int number of failures (non-negative integers; can include zero)
#' @param .time:logical logical; is case time-censored data?
#' @param .failure:logical logical; is case failure-censored data?
dk = function(x, r, .time = FALSE, .failure = FALSE){
  # Testing values
  # x = 2; r = 20; .time = FALSE; .failure = FALSE

  # The k-factor equals chisq(df) / (2*r), so its density has the exact closed
  # form 2*r * dchisq(2*r*x, df) (and is 0 for x < 0, since the k-factor >= 0).
  # df follows the same rule as qk(): 2*r for complete or Type II
  # (failure-censored) data, 2*(r + 1) for time-censored data.

  # Input error handling (mirror qk())
  stopifnot(is.numeric(x))
  stopifnot(is.numeric(as.integer(r)))
  stopifnot(is.logical(.time))
  stopifnot(is.logical(.failure))
  stopifnot(
    (.time == TRUE & .failure == FALSE) |
      (.time == FALSE & .failure == FALSE) |
      (.time == FALSE & .failure == TRUE) )
  stopifnot( (r > 0) | (r == 0 & .time == TRUE) )

  # Zero failures (time-censored): the k-factor is Exponential(rate = 1)
  if(r == 0){ return(dexp(x, rate = 1)) }

  # Degrees of freedom for the chi-squared numerator of the k-factor
  df = if(.time == TRUE){ 2 * (r + 1) }else{ 2 * r }
  # Exact density
  d = 2 * r * dchisq(2 * r * x, df = df)
  return(d)
}
# Example
# dk(x = c(0, 1, 2, 3), r = 21, .time = TRUE, .failure = FALSE)

File: functions/functions_k.py
# functions_k.py

"""
k-factor Functions for Python

This module provides Python equivalents of the k-factor functions from functions_k.R.
These functions are used for estimating confidence intervals and visualizing 
sampling distributions of failure rates.
"""

from scipy.stats import chi2
from numpy import log, array, isnan
from pandas import Series
import warnings

# Import utility functions from functions_distributions
from functions_distributions import runif, approxfun, density, tidy_density, seq


def qk(p, r, time=False, failure=False):
    """
    k-factor Quantiles
    
    Function to return quantiles for k-factors.
    Intended for estimating confidence intervals for failure rates.
    
    Parameters
    ----------
    p : array-like
        Vector of probabilities / percentile(s) (must be between 0 and 1)
    r : int
        Number of failures (non-negative integers; can include zero)
    time : bool, optional
        Logical; is case time-censored data? (default: False)
    failure : bool, optional
        Logical; is case failure-censored data? (default: False)
    
    Returns
    -------
    Series
        k-factor quantiles corresponding to the input probabilities
    
    Notes
    -----
    - time and failure cannot both be True
    - If r == 0, then time must be True
    - Returns NA values for invalid input combinations with a warning message
    """
    # Convert input to array for vectorized operations
    p = array(p)
    r = int(r)
    
    # Input error handling
    if not all((p >= 0) & (p <= 1)):
        raise ValueError("All probabilities p must be between 0 and 1")
    
    if not isinstance(r, int) or r < 0:
        raise ValueError("r must be a non-negative integer")
    
    if not isinstance(time, bool):
        raise ValueError("time must be a boolean")
    
    if not isinstance(failure, bool):
        raise ValueError("failure must be a boolean")
    
    # Check mutual exclusivity: time and failure cannot both be True
    if time and failure:
        raise ValueError("time and failure cannot both be True")
    
    # Check zero-failure condition
    if r == 0 and not time:
        raise ValueError("If r == 0, then time must be True")
    
    if r == 0 and failure:
        raise ValueError("If r == 0, then failure cannot be True")
    
    # Evaluate if p is in the upper or lower tail
    upper = p > 0.5
    
    # Does r == 0?
    zerofailures = (r == 0)
    
    # Initialize result array with NaN
    k = array([float('nan')] * len(p))
    
    # Apply conditional logic (equivalent to R's case_when)
    # Convert boolean arrays to conditions for each element
    for i in range(len(p)):
        if not zerofailures and not time and not failure and upper[i]:
            # 1+ failures AND complete data AND UPPER tail --> Get k-factor for r as normal
            k[i] = chi2.ppf(p[i], df=2*r) / (2*r)
        elif not zerofailures and not time and not failure and not upper[i]:
            # 1+ failures AND complete data AND LOWER tail --> Get k-factor for r as normal
            k[i] = chi2.ppf(p[i], df=2*r) / (2*r)
        elif not zerofailures and time and not failure and upper[i]:
            # 1+ failures AND time-censored data AND UPPER tail --> Get k-factor for r+1
            k[i] = chi2.ppf(p[i], df=2*(r + 1)) / (2*r)
        elif not zerofailures and time and not failure and not upper[i]:
            # 1+ failures AND time-censored data AND LOWER tail --> Get k-factor for r as normal
            k[i] = chi2.ppf(p[i], df=2*r) / (2*r)
        elif not zerofailures and not time and failure and upper[i]:
            # 1+ failures AND failure-censored data AND UPPER tail --> Get k-factor with adjustment
            k[i] = chi2.ppf(p[i], df=2*((r-1) + 1)) / (2 * (r-1)) * (r-1)/r
        elif not zerofailures and not time and failure and not upper[i]:
            # 1+ failures AND failure-censored data AND LOWER tail --> Get k-factor with r as normal
            k[i] = chi2.ppf(p[i], df=2*r) / (2*r)
        elif zerofailures and time and not failure:
            # If zero failures --> then time-censored --> time = TRUE, and upper/lower distinction doesn't matter.
            k[i] = -log(1 - p[i])
        else:
            # Otherwise, return NA.
            k[i] = float('nan')
    
    # Check for NA values and warn
    if any(isnan(k)):
        warnings.warn("At least 1 k-factor could not be calculated, due to improper inputs. Review the rules for time-censored, failure-censored, and zero-failure data.")
    
    return Series(k)


def rk(n, r, time=False, failure=False):
    """
    k-factor Random Deviates
    
    Get a random sample of k-factor values for simulating sampling distributions 
    of failure rates.
    
    Parameters
    ----------
    n : int
        Number of observations (must be positive integer)
    r : int
        Number of failures (non-negative integers; can include zero)
    time : bool, optional
        Logical; is case time-censored data? (default: False)
    failure : bool, optional
        Logical; is case failure-censored data? (default: False)
    
    Returns
    -------
    Series
        Random sample of k-factor values
    """
    # Input error handling
    n = int(n)
    if n <= 0:
        raise ValueError("n must be a positive integer")
    
    if not isinstance(time, bool):
        raise ValueError("time must be a boolean")
    
    if not isinstance(failure, bool):
        raise ValueError("failure must be a boolean")
    
    # Check mutual exclusivity
    if time and failure:
        raise ValueError("time and failure cannot both be True")
    
    # Generate a uniform distribution of percentiles p
    p_uniform = runif(n=n, min=0, max=1)
    
    # Return quantiles for the random percentiles
    k = qk(p=p_uniform, r=r, time=time, failure=failure)
    return k


def pk(q, r, time=False, failure=False):
    """
    k-factor Cumulative Distribution Function
    
    Function to return cumulative probabilities / percentiles given a supplied 
    k-factor quantile `q`.
    Intended for confidence intervals for failure rates.
    
    Parameters
    ----------
    q : array-like
        Vector of quantiles (k-factors)
    r : int
        Number of failures (non-negative integers; can include zero)
    time : bool, optional
        Logical; is case time-censored data? (default: False)
    failure : bool, optional
        Logical; is case failure-censored data? (default: False)
    
    Returns
    -------
    Series
        Cumulative probabilities corresponding to the input quantiles
    """
    # Convert input to array
    q = array(q)
    
    # Construct an approximation function f, which gives the inverse of the Quantile Function,
    # such that you use linear interpolation to return a Probability for any Quantile supplied.
    by = 0.001
    p_range_list = [by/10000, by/1000, by/100, by/10]
    p_range_list.extend(seq(from_=0, to=1, by=by).tolist())
    p_range_list.extend([1 - by/10, 1 - by/100, 1 - by/1000, 1 - by/10000])
    p_range = sorted(p_range_list)
    
    # Get the quantiles for that range
    q_range = qk(p=p_range, r=r, time=time, failure=failure)
    
    # Remove NaN values for interpolation
    valid_mask = ~q_range.isna()
    q_range_clean = q_range[valid_mask]
    p_range_clean = [p_range[i] for i in range(len(p_range)) if valid_mask.iloc[i]]
    
    # Create DataFrame for approxfun (it expects x and y columns)
    from pandas import DataFrame
    data = DataFrame({'x': q_range_clean, 'y': p_range_clean})
    
    # Get the inverse quantile function
    # Note: R's approxfun with rule=2 means extrapolate, which is the default in Python's interp1d
    f = approxfun(data, fill_value='extrapolate', bounds_error=False)
    
    # Return the expected CDF for that quantile
    p = f(q)
    
    # Convert to Series if not already
    if not isinstance(p, Series):
        p = Series(p)
    
    return p


def dk(x, r, time=False, failure=False):
    """
    k-factor Probability Density Function
    
    Function to return probability densities given a supplied k-factor quantile `x`.
    Intended for visualizing sampling distributions of failure rates.
    
    Parameters
    ----------
    x : array-like
        Vector of quantiles (k-factors)
    r : int
        Number of failures (non-negative integers; can include zero)
    time : bool, optional
        Logical; is case time-censored data? (default: False)
    failure : bool, optional
        Logical; is case failure-censored data? (default: False)
    
    Returns
    -------
    Series
        Probability densities corresponding to the input quantiles
    """
    # Convert input to array
    x = array(x)
    
    # Construct a range of quantiles corresponding to a range of cumulative probabilities
    by = 0.001
    p_range_list = [by/10000, by/1000, by/100, by/10]
    p_range_list.extend(seq(from_=0, to=1, by=by).tolist())
    p_range_list.extend([1 - by/10, 1 - by/100, 1 - by/1000, 1 - by/10000])
    p_range = sorted(p_range_list)
    
    # Get the quantiles for that range
    q_range = qk(p=p_range, r=r, time=time, failure=failure)
    
    # Filter out NaN values for density estimation
    q_range_clean = q_range.dropna()
    
    # Fit a density curve to that quantile data
    # Note: R's density() with cut=c(0) truncates at 0, but Python's gaussian_kde
    # doesn't have this option. We'll use the standard density() function.
    model = density(q_range_clean)
    
    # Get tidy density DataFrame with x and y columns
    density_df = tidy_density(model, n=1000)
    
    # Approximate a function using the density curve's x and y values
    f = approxfun(density_df, fill_value='extrapolate', bounds_error=False)
    
    # Estimate density
    d = f(x)
    
    # Convert to Series if not already
    if not isinstance(d, Series):
        d = Series(d)
    
    return d

